cpp/oneapi/dal/backend/memory.hpp

```diff
@@ -16,37 +16,117 @@
 
 #pragma once
 
-#ifdef ONEDAL_DATA_PARALLEL
-#include <CL/sycl.hpp>
-#endif
-
-#include "oneapi/dal/detail/common.hpp"
+#include "oneapi/dal/array.hpp"
+#include "oneapi/dal/backend/common.hpp"
 
 namespace oneapi::dal::backend {
 
 #ifdef ONEDAL_DATA_PARALLEL
+inline bool is_device_usm_pointer(const sycl::queue& queue, const void* pointer) {
+    const auto pointer_type = sycl::get_pointer_type(pointer, queue.get_context());
+    return pointer_type == sycl::usm::alloc::device;
+}
+
+inline bool is_shared_usm_pointer(const sycl::queue& queue, const void* pointer) {
+    const auto pointer_type = sycl::get_pointer_type(pointer, queue.get_context());
+    return pointer_type == sycl::usm::alloc::shared;
+}
+
+inline bool is_host_usm_pointer(const sycl::queue& queue, const void* pointer) {
+    const auto pointer_type = sycl::get_pointer_type(pointer, queue.get_context());
+    return pointer_type == sycl::usm::alloc::host;
+}
```

Comment:
The reason to create such wrappers? Even you do not use it in `is_device_friendly_usm_pointer` by yourself.

Reply:
The reason is to not write every time something like this:
```c++
ONEDAL_ASSERT(sycl::get_pointer_type(ary.get_data(), queue.get_context()) == sycl::usm::alloc::host);
```

But write this:
```c++
ONEDAL_ASSERT(is_host_usm_pointer(queue, ary.get_data());
```

>  Even you do not use it in is_device_friendly_usm_pointer by yourself.

I'm trying to minimize number of calls to `get_pointer_type`, maybe it's overoptimization.